pkg/peeker/reader.go

```diff
@@ -37,17 +39,23 @@ func (r *Reader) Limit() int {
 	return r.limit
 }
 
-func (r *Reader) fill(min int) error {
-	if min > r.limit {
+func (r *Reader) fill(need int) error {
+	if need > r.limit {
 		return ErrBufferOverflow
 	}
-	if min > cap(r.buffer) {
-		r.buffer = make([]byte, min)
+	if need > cap(r.buffer) {
+		r.buffer = make([]byte, need)
 	}
 	r.buffer = r.buffer[:cap(r.buffer)]
 	copy(r.buffer, r.cursor)
 	clen := len(r.cursor)
-	n, err := io.ReadAtLeast(r.Reader, r.buffer[clen:], min)
+	var request int
+	if r.interactive {
+		request = need - clen
+	} else {
+		request = len(r.buffer) - clen
+	}
+	n, err := io.ReadAtLeast(r.Reader, r.buffer[clen:], request)
```

Comment:
I don't think this path offers any benefit. `io.ReadAtLeast`'s first call to `r.Reader.Read` will fill `r.buffer[clen:]` if enough bytes are available (as they usually are when reading from a file or object store). But if it doesn't, why block here until it's filled instead of allowing the caller to make forward progress once `min` bytes are available?

Reply:
The effect here isn't as pronounced for ZNG now that we have frames but the behavior for the old scanner (and for possible future users of Peeker... this is in pkg) will do a system call for every small read when reading from a reader like the file system that otherwise doesn't run asynchronously.  In terms of letting the reader make forward progress, pipelining will work well reading a large buffer at a time as the reader issues the next big read and the scanners are busy working on the previous buffer.  If reading is the bottleneck, doing lots of small reads to let the scanners make forward progress doesn't solve the problem.

Reply:
Maybe I'm missing something.  Is it that the read is filling the buffer when data is available beyond the request size but waiting only until the min is satisfied?

Reply:
> Is it that the read is filling the buffer when data is available beyond the request size but waiting only until the min is satisfied?

That's correct. `io.Reader.Read` semantics are conventionally "give me as much as you can without blocking." The only thing `io.ReadAtLeast` adds to that is a loop that calls `Read` until it has at least `min` bytes.

You can see this if you put something like this right after the call to `io.ReadAtLeast` and then read a large file with `zq`. The first few and last few reads will be short but all the rest will be full-length.
```go
	if n < len(r.buffer[clen:]) {
		fmt.Println("short", n, len(r.buffer[clen:]))
	} else {
		fmt.Println("full", n)
	}
```